File: txtscrio/InputStr.c
// This file is part of Adlib Tracker II (AT2).
//
// SPDX-FileType: SOURCE
// SPDX-FileCopyrightText: 2014-2024 The Adlib Tracker 2 Authors
// SPDX-License-Identifier: GPL-3.0-or-later

String_t InputStr (const String *s, uint8_t x, uint8_t y, uint8_t ln,
                   uint8_t ln1, uint8_t attr1, uint8_t attr2) {

  String_t r, s1, s2;
  bool appn, for1st, qflg, chflag, ins;
  uint8_t cloc, xloc, xint, attr;
  uint16_t key;

#if GO32
  DBG_ENTER ("InputStr");
#endif // GO32

  r = Copy (s, 1, ln);

  if (is_environment.locate_pos > ln1)
    is_environment.locate_pos = ln1;
  if (is_environment.locate_pos > r.len + 1)
    is_environment.locate_pos = r.len;

  cloc = is_environment.locate_pos;
  xloc = is_environment.locate_pos;
  xint = x;
  qflg = false;
  ins  = is_setting.insert_mode;
  appn = !is_setting.append_enabled;

  x--;

  if (ins) {
    if (use_large_cursor)
      WideCursor ();
    else
      ThinCursor ();
  } else {
    if (use_large_cursor)
      ThinCursor ();
    else
      WideCursor ();
  }

  CopyString ((String *) &s1, (String *) &r, 255);
  if (s1.len > ln1)
    s1.len = ln1;

  {
    String_t t;

    t = ExpStrR ((String *) "", ln1, ' ');
    ShowStr (screen_ptr, xint, y, (String *) &t, attr1);

    t = FilterStr2 ((String *) &s1, is_setting.char_filter, '_');
    ShowStr (screen_ptr, xint, y, (String *) &t, attr2);
  }

  for1st = true;

  do {
    CopyString ((String *) &s2, (String *) &s1, 255);
    s1 = Copy ((String *) &r, cloc - (xloc - 1), ln1);

    attr = appn ? attr1 : attr2;

    if (appn && for1st) {
      String_t t, t1;

      t = FilterStr2 ((String *) &s1, is_setting.char_filter, '_');
      t1 = ExpStrR ((String *) &t, ln1, ' ');
      ShowStr (screen_ptr, xint, y, (String *) &t1, attr1);
      for1st = false;
    }

    if ((s1.len != s2.len)
    ||  (strncmp ((char *) s1.str, (char *) s2.str,
                  (s1.len <= s2.len) ? s1.len : s2.len))) {
      String_t t, t1;

      t = FilterStr2 ((String *) &s1, is_setting.char_filter, '_');
      t1 = ExpStrR ((String *) &t, ln1, ' ');
      ShowStr (screen_ptr, xint, y, (String *) &t1, attr1);
    }

    if (ln1 < ln) { // FIXME: (lint) this condition is the same as next
      uint8_t t[2];
      uint8_t a;

      if (r.len && (cloc > xloc)) {
        t[0] = 1;
        t[1] = '\x11'; // HINT: (AT2) custom glyph
        a = (attr & 0xF0) + 0x0F;
      } else {
        if (r.len && (cloc == xloc)) {
          t[0] = 1;
          t[1] = r.str[0];
          a = attr;
        } else {
          t[0] = 1;
          t[1] = ' ';
          a = attr1;
        }
      }
      ShowStr (screen_ptr, xint, y, (String *) &t, a);
    }

    if (ln1 < ln) { // FIXME: (lint) this condition is the same as previous
      String_t t;
      uint8_t a;

      if (cloc - xloc + ln1 < r.len) {
        t.len = 1;
        t.str[0] = '\x10'; // HINT: (AT2) custom glyph
        a = (attr & 0xF0) + 0x0F;
      } else {
        if (cloc - xloc + ln1 == r.len) {
          uint8_t t1[2];

          t1[0] = 1;
          t1[1] = r.str[r.len - 1];
          t = FilterStr2 ((String *) &t1, is_setting.char_filter, '_');
          a = attr;
        } else {
          t.len = 1;
          t.str[0] = ' ';
          a = attr1;
        }
      }
      ShowStr (screen_ptr, xint + ln1 - 1, y, (String *) &t, a);
    }

    GotoXY (x + xloc, y);

    if (keypressed ()) {
      key = getkey ();

      qflg = qflg || LookUpKey (key, is_setting.terminate_keys,
                                sizeof (is_setting.terminate_keys)
                                / sizeof (is_setting.terminate_keys[0]));

      chflag = false;

      if (!qflg) {
        switch (key) {
        case kTAB:
          appn = true;
          break;

        case kCHplus:
        case kNPplus:
          chflag = true;
          {
            uint_least8_t base;

            if (memcmp (is_setting.character_set,
                        DEC_NUM_CHARSET, sizeof (CharSet_t)) == 0)
              base = 10;
            else if (memcmp (is_setting.character_set,
                             HEX_NUM_CHARSET, sizeof (CharSet_t)) == 0)
              base = 16;
            else
              base = 0;

            if (base) {
              int64_t v = Str2num ((String *) &r, base) + 1;

              if (v <= is_environment.max_num) {
                String_t t = Num2str (v, base);

                if (t.len <= ln1) {
                  CopyString ((String *) &r, (String *) &t, 255);
                  chflag = false;
                }
              }
            }
          }
          break;

        case kCHmins:
        case kNPmins:
          {
            uint_least8_t base;

            if (memcmp (is_setting.character_set,
                        DEC_NUM_CHARSET, sizeof (CharSet_t)) == 0)
              base = 10;
            else if (memcmp (is_setting.character_set,
                             HEX_NUM_CHARSET, sizeof (CharSet_t)) == 0)
              base = 16;
            else
              base = 0;

            if (base) {
              int64_t v = Str2num ((String *) &r, base) - 1;

              if ((v >= 0) && (v >= is_environment.min_num)) {
                String_t t = Num2str (v, base);

                if (t.len <= ln1) {
                  CopyString ((String *) &r, (String *) &t, 255);
                  chflag = false;
                }
              }
            }
          }
          break;

        case kCtrlY:
          appn = true;
          r.len = 0;
          cloc = 1;
          xloc = 1;
          break;

        case kCtrlT:
          appn = true;

          while ((cloc <= r.len)
          &&     bit_test (is_setting.word_characters, r.str[cloc - 1]))
            Delete ((String *) &r, cloc, 1);

          while ((cloc <= r.len)
          &&     !bit_test (is_setting.word_characters, r.str[cloc - 1]))
            Delete ((String *) &r, cloc, 1);
          break;

        case kCtrlK:
          appn = true;
          Delete ((String *) &r, cloc, r.len);
          break;

        case kCtBkSp:
          appn = true;

          while ((cloc > 1)
          &&     bit_test (is_setting.word_characters, r.str[cloc - 1 - 1])) {
            cloc--;
            Delete ((String *) &r, cloc, 1);
            if (xloc > 1)
              xloc--;
          }

          while ((cloc > 1)
          &&     !bit_test (is_setting.word_characters, r.str[cloc - 1])) {
            cloc--;
            Delete ((String *) &r, cloc, 1);
            if (xloc > 1)
              xloc--;
          }
          break;

        case kBkSPC:
          appn = true;
          if (cloc > 1) {
            cloc--;
            Delete ((String *) &r, cloc, 1);
            if (xloc > 1)
              xloc--;
          }
          break;

        case kDELETE:
          appn = true;
          if (cloc <= r.len)
            Delete ((String *) &r, cloc, 1);
          break;

        case kCtLEFT:
          appn = true;

          while ((cloc > 1)
          &&     bit_test (is_setting.word_characters, r.str[cloc - 1])) {
            cloc--;
            if (xloc > 1)
              xloc--;
          }

          while ((cloc > 1)
          &&     !bit_test (is_setting.word_characters, r.str[cloc - 1])) {
            cloc--;
            if (xloc > 1)
              xloc--;
          }
          break;

        case kCtRGHT:
          appn = true;

          while ((cloc < r.len)
          &&     bit_test (is_setting.word_characters, r.str[cloc - 1])) {
            cloc++;
            if (xloc < ln1)
              xloc++;
          }

          while ((cloc < r.len)
          &&     !bit_test (is_setting.word_characters, r.str[cloc - 1])) {
            cloc++;
            if (xloc < ln1)
              xloc++;
          }
          break;

        case kLEFT:
          appn = true;
          if (cloc > 1)
            cloc--;
          if (xloc > 1)
            xloc--;
          break;

        case kRIGHT:
          appn = true;
          if ((cloc < r.len)
          ||  ((cloc == r.len) && ((r.len < ln) || (r.len < ln1))))
            cloc++;
          if ((xloc <= r.len) && (xloc < ln1))
            xloc++;
          break;

        case kINSERT:
          if (is_setting.replace_enabled) {
            ins = !ins;
            if (ins) {
              if (use_large_cursor)
                WideCursor ();
              else
                ThinCursor ();
            } else {
              if (use_large_cursor)
                ThinCursor ();
              else
                WideCursor ();
            }
          }
          break;

        case kHOME:
          appn = true;
          cloc = 1;
          xloc = 1;
          break;

        case kEND:
          appn = true;
          cloc = r.len + (((r.len < ln) || (r.len < ln1)) ? 1 : 0);
          xloc = (cloc < ln1) ? cloc : ln1;
          break;

        default:
          chflag = true;
          break;
        }
      }

      if (chflag && bit_test (is_setting.character_set, key & 0xFF)) {
        String_t t;

        if (!appn) {
          r.len = 0;
          cloc = 1;
          xloc = 1;
        }
        appn = true;

        t = CutStrR ((String *) &r, cloc);
        if (ins && (t.len < ln)) {
          if (t.len < ln) // FIXME: (lint) this condition is always true
            InsertChr (key & 0xFF, (String *) &r, 255, cloc);
          else
            r.str[cloc - 1] = key & 0xFF;

          t = FilterStr2 ((String *) &r, is_setting.valid_chars, '_');
          CopyString ((String *) &r, (String *) &t, 255);
          if (cloc < ln)
            cloc++;
          if ((xloc < ln) && (xloc < ln1))
            xloc++;
        } else if ((r.len < ln) || !ins) {
          if ((cloc > r.len) && (r.len < ln))
            r.len++;
          r.str[cloc - 1] = key & 0xFF;
          t = FilterStr2 ((String *) &r, is_setting.valid_chars, '_');
          CopyString ((String *) &r, (String *) &t, 255);
          if (cloc < ln)
            cloc++;
          if ((xloc < ln) && (xloc < ln1))
            xloc++;
        }
      }
    }

    CopyString (is_environment.cur_str, (String *) &r, 255);
    if (is_environment.ext_proc)
      is_environment.ext_proc ();
#if GO32
    //draw_screen ();
    keyboard_reset_buffer_alt ();
#else // !GO32
    draw_screen ();
    //keyboard_reset_buffer ();
#endif // !GO32
  } while (!qflg);

  is_environment.locate_pos = cloc ? cloc : 1;
  is_environment.keystroke = key;
  is_environment.insert_mode = ins;

  return r;
}

File: txtscrio/pas/InputStr.pas
// This file is part of Adlib Tracker II (AT2).
//
// SPDX-FileType: SOURCE
// SPDX-FileCopyrightText: 2014-2024 The Adlib Tracker 2 Authors
// SPDX-License-Identifier: GPL-3.0-or-later

function InputStr(s: String; x,y,ln,ln1: Byte; atr1,atr2: Byte): String;

var
  appn,for1st,qflg,chflag,ins: Boolean;
  cloc,xloc,xint,attr: Byte;
  key: Word;
  s1,s2: String;

function more(value1,value2: Byte): Byte;
begin
  If (value1 >= value2) then more := value1
  else more := value2;
end;

label _end;

begin { InputStr }
  _dbg_enter ({$I %FILE%}, 'InputStr');

  s := Copy(s,1,ln);
  If (is_environment.locate_pos > ln1) then
    is_environment.locate_pos := ln1;
  If (is_environment.locate_pos > Length(s)+1) then
    is_environment.locate_pos := Length(s);

  cloc := is_environment.locate_pos;
  xloc := is_environment.locate_pos;
  xint := x;
  qflg := FALSE;
  ins  := is_setting.insert_mode;
  appn := NOT is_setting.append_enabled;

  Dec(x);
  If ins then
    If use_large_cursor then WideCursor
    else ThinCursor
  else
    If use_large_cursor then ThinCursor
    else WideCursor;

  s1 := s;
  If (BYTE(s1[0]) > ln1) then s1[0] := CHR(ln1);

  ShowStr(screen_ptr,xint,y,ExpStrR('',ln1,' '),atr1);
  ShowStr(screen_ptr,xint,y,FilterStr2(s1,is_setting.char_filter,'_'),atr2);
  for1st := TRUE;

  Repeat
    s2 := s1;
    If (xloc = 1) then s1 := Copy(s,cloc,ln1)
    else s1 := Copy(s,cloc-xloc+1,ln1);

    If NOT appn then attr := atr2
    else attr := atr1;

    If appn and for1st then
      begin
        ShowStr(screen_ptr,xint,y,ExpStrR(FilterStr2(s1,is_setting.char_filter,'_'),ln1,' '),atr1);
        for1st := FALSE;
      end;

    If (s2 <> s1) then
      ShowStr(screen_ptr,xint,y,ExpStrR(FilterStr2(s1,is_setting.char_filter,'_'),ln1,' '),atr1);

    If (ln1 < ln) then
      If (cloc-xloc > 0) and (Length(s) > 0) then
        ShowStr(screen_ptr,xint,y,#17,(attr AND $0f0)+$0f)
      else If (cloc-xloc = 0) and (Length(s) <> 0) then
             ShowStr(screen_ptr,xint,y,s[1],attr)
           else
             ShowStr(screen_ptr,xint,y,' ',atr1);

    If (ln1 < ln) then
      If (cloc-xloc+ln1 < Length(s)) then
        ShowStr(screen_ptr,xint+ln1-1,y,#16,(attr AND $0f0)+$0f)
      else If (cloc-xloc+ln1 = Length(s)) then
             ShowStr(screen_ptr,xint+ln1-1,y,FilterStr2(s[Length(s)],is_setting.char_filter,'_'),attr)
           else
             ShowStr(screen_ptr,xint+ln1-1,y,' ',atr1);

    GotoXY(x+xloc,y);
    If keypressed then key := GetKey else GOTO _end;
    If LookupKey(key,is_setting.terminate_keys,50) then qflg := TRUE;
    chflag := FALSE;

    If NOT qflg then
      Case key of
        kTAB: appn := TRUE;

        kCHplus,
        kNPplus: begin
                   chflag := TRUE;
                   If (is_setting.character_set = DEC_NUM_CHARSET) then
                     If (Length(Num2str(SUCC(Str2num(s,10)),10)) <= ln1) and
                        (SUCC(Str2num(s,10)) <= is_environment.max_num) then
                       begin
                         s := Num2str(SUCC(Str2num(s,10)),10);
                         chflag := FALSE;
                       end
                     else
                   else If (is_setting.character_set = HEX_NUM_CHARSET) then
                          If (Length(Num2str(SUCC(Str2num(s,16)),16)) <= ln1) and
                             (SUCC(Str2num(s,16)) <= is_environment.max_num) then
                            begin
                              s := Num2str(SUCC(Str2num(s,16)),16);
                              chflag := FALSE;
                            end;
                 end;
        kCHmins,
        kNPmins: begin
                   chflag := TRUE;
                   If (is_setting.character_set = DEC_NUM_CHARSET) then
                     If (Str2num(s,10) > 0) and
                        (PRED(Str2num(s,10)) >= is_environment.min_num) then
                       begin
                         s := Num2str(PRED(Str2num(s,10)),10);
                         chflag := FALSE;
                       end
                     else
                   else If (is_setting.character_set = HEX_NUM_CHARSET) then
                          If (Str2num(s,16) > 0) and
                             (PRED(Str2num(s,16)) >= is_environment.min_num) then
                            begin
                              s := Num2str(PRED(Str2num(s,16)),16);
                              chflag := FALSE;
                            end;
                end;

        kCtrlY: begin
                  appn := TRUE;
                  s := '';
                  cloc := 1;
                  xloc := 1;
                end;

        kCtrlT: begin
                  appn := TRUE;
                  While (s[cloc] in is_setting.word_characters) and
                        (cloc <= Length(s)) do Delete(s,cloc,1);

                  While NOT (s[cloc] in is_setting.word_characters) and
                            (cloc <= Length(s)) do Delete(s,cloc,1);
                end;

        kCtrlK: begin
                  appn := TRUE;
                  Delete(s,cloc,Length(s));
                end;

        kCtBkSp: begin
                   appn := TRUE;
                   While (s[cloc-1] in is_setting.word_characters) and
                         (cloc > 1) do
                     begin
                       Dec(cloc); Delete(s,cloc,1);
                       If (xloc > 1) then Dec(xloc);
                     end;

                   While NOT (s[cloc-1] in is_setting.word_characters) and
                             (cloc > 1) do
                     begin
                       Dec(cloc); Delete(s,cloc,1);
                       If (xloc > 1) then Dec(xloc);
                     end;
                 end;

        kBkSPC: begin
                  appn := TRUE;
                  If (cloc > 1) then
                    begin
                      If (xloc > 1) then Dec(xloc);
                      Dec(cloc); Delete(s,cloc,1);
                    end;
                end;

        kDELETE: begin
                   appn := TRUE;
                   If (cloc <= Length(s)) then Delete(s,cloc,1);
                 end;

        kCtLEFT: begin
                   appn := TRUE;
                   While (s[cloc] in is_setting.word_characters) and
                         (cloc > 1) do
                     begin
                       Dec(cloc);
                       If (xloc > 1) then Dec(xloc);
                     end;

                   While NOT (s[cloc] in is_setting.word_characters) and
                             (cloc > 1) do
                     begin
                       Dec(cloc);
                       If (xloc > 1) then Dec(xloc);
                     end;
                 end;

        kCtRGHT: begin
                   appn := TRUE;
                   While (s[cloc] in is_setting.word_characters) and
                         (cloc < Length(s)) do
                     begin
                       Inc(cloc);
                       If (xloc < ln1) then Inc(xloc);
                     end;

                   While NOT (s[cloc] in is_setting.word_characters) and
                             (cloc < Length(s)) do
                     begin
                       Inc(cloc);
                       If (xloc < ln1) then Inc(xloc);
                     end;
                 end;

        kLEFT: begin
                 appn := TRUE;
                 If (cloc > 1) then Dec(cloc);
                 If (xloc > 1) then Dec(xloc);
               end;

        kRIGHT: begin
                  appn := TRUE;
                  If (cloc < Length(s)) or ((cloc = Length(s)) and
                       ((Length(s) < more(ln,ln1)))) then
                    Inc(cloc);
                  If (xloc < ln1) and (xloc <= Length(s)) then Inc(xloc);
                end;

        kINSERT: If is_setting.replace_enabled then
                   begin
                     ins := NOT ins;
                     If ins then
                       If use_large_cursor then WideCursor
                       else ThinCursor
                     else
                       If use_large_cursor then ThinCursor
                       else WideCursor;
                   end;

        kHOME: begin
                 appn := TRUE;
                 cloc := 1;
                 xloc := 1;
               end;

        kEND: begin
                appn := TRUE;
                If (Length(s) < more(ln,ln1)) then cloc := Succ(Length(s))
                else cloc := Length(s);
                If (cloc < ln1) then xloc := cloc else xloc := ln1;
              end;

        else chflag := TRUE;
      end;

    If chflag and (CHR(LO(key)) in tCHARSET(is_setting.character_set)) then
      begin
        If NOT appn then begin s := ''; cloc := 1; xloc := 1; end;
        appn := TRUE;
        If ins and (Length(CutStrR(s,cloc)) < ln) then
          begin
            If (Length(CutStrR(s,cloc)) < ln) then
              Insert(CHR(LO(key)),s,cloc)
            else s[cloc] := CHR(LO(key));
            s := FilterStr2(s,is_setting.valid_chars,'_');
            If (cloc < ln) then Inc(cloc);
            If (xloc < ln) and (xloc < ln1) then Inc(xloc)
          end
        else
          If (Length(s) < ln) or NOT ins then
            begin
              If (cloc > Length(s)) and (Length(s) < ln) then
                Inc(BYTE(s[0]));
              s[cloc] := CHR(LO(key));
              s := FilterStr2(s,is_setting.valid_chars,'_');
              If (cloc < ln) then Inc(cloc);
              If (xloc < ln) and (xloc < ln1) then Inc(xloc);
            end;
      end;
_end:
      is_environment.cur_str := s;
      If (Addr(is_environment.ext_proc) <> NIL) then is_environment.ext_proc;
{$IFDEF GO32V2}
      // draw_screen;
      keyboard_reset_buffer_alt;
{$ELSE}
      draw_screen;
      // keyboard_reset_buffer;
{$ENDIF}
  until qflg;

  If (cloc = 0) then is_environment.locate_pos := 1
  else is_environment.locate_pos := cloc;
  is_environment.keystroke := key;
  is_environment.insert_mode := ins;
  InputStr := s;

  _dbg_leave; //EXIT //InputStr
end;
